contracts/BrandedToken.sol

```diff
@@ -0,0 +1,566 @@
+     *      Constructor requires:
+     *          - valueToken address is not zero
+     *          - conversionRate is not zero
+     *
+     * @param _valueToken The value to which valueToken is set.
+     * @param _symbol The value to which tokenSymbol, defined in EIP20Token,
+     *                is set.
+     * @param _name The value to which tokenName, defined in EIP20Token,
+     *              is set.
+     * @param _decimals The value to which tokenDecimals, defined in EIP20Token,
+     *                  is set.
+     * @param _conversionRate The value to which conversionRate is set.
+     * @param _conversionRateDecimals The value to which
+     *                                conversionRateDecimals is set.
+     * @param _organization The value to which organization, defined in Organized,
+     *                      is set.
+     */
+    constructor(
+        EIP20Interface _valueToken,
+        string memory _symbol,
+        string memory _name,
+        uint8 _decimals,
+        uint256 _conversionRate,
+        uint8 _conversionRateDecimals,
+        OrganizationInterface _organization
+    )
+        EIP20Token(_symbol, _name, _decimals)
+        Organized(_organization)
+        public
+    {
+        require(
+            address(_valueToken) != address(0),
+            "ValueToken is zero."
+        );
+        require(
+            _conversionRate != 0,
+            "ConversionRate is zero."
+        );
+
+        valueToken = _valueToken;
+        conversionRate = _conversionRate;
+        conversionRateDecimals = _conversionRateDecimals;
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Transfers value tokens from msg.sender to itself,
+     *         stores the amount of branded tokens to mint if request
+     *         is accepted, and emits stake request information.
+     *
+     * @dev It is expected that this contract will have a sufficientallowance
+     *      to transfer value tokens from the staker at the time this function
+     *      is executed.
+     *
+     *      Function requires:
+     *          - _mint is equivalent to _stake
+     *          - msg.sender does not have a stake request hash
+     *          - valueToken.transferFrom returns true
+     *
+     * @param _stake Amount of value tokens to stake.
+     * @param _mint Amount of branded tokens to mint.
+     *
+     * @return stakeRequestHash_ Hash of stake request information calculated per
+     *                           EIP 712.
+     */
+    function requestStake(
+        uint256 _stake,
+        uint256 _mint
+    )
+        external
+        returns (bytes32 stakeRequestHash_)
+    {
+        require(
+            _mint == convertToBrandedTokens(_stake),
+            "Mint is not equivalent to stake."
+        );
+        require(
+            stakeRequestHashes[msg.sender] == "",
+            "Staker has a stake request hash."
+        );
+
+        // TODO: confirm alignment with nonce updating in Gateway
```

Comment:
in gateway the nonce starts at 0 and is updated after the action;

Reply:
Thanks! Updated.